src/sc/sidechain.cpp

```diff
@@ -606,36 +947,179 @@ bool ScMgr::erase(const uint256& scId)
 bool ScMgr::sidechainExists(const uint256& scId) const
 {
     LOCK(sc_lock);
-    return ManagerScInfoMap.count(scId);
+    return mManagerScInfoMap.count(scId);
+}
+
+int ScMgr::getCertificateMaxIncomingHeight(const uint256& scId, int epochNumber)
+{
+    ScInfo info;
+    if (!instance().getScInfo(scId, info) )
+    {
+        LogPrint("cert", "%s():%d - scId[%s] not found, returning -1\n", __func__, __LINE__, scId.ToString() );
+        return -1;
+    }
+
+    // the safety margin from the end of referred epoch is computed as 20% of epoch length + 1
+    // TODO move this in consensus params
+    int val = info.creationBlockHeight + (epochNumber * info.creationData.withdrawalEpochLength) +
+        (int)(info.creationData.withdrawalEpochLength/5) + 1;
+
+    LogPrint("cert", "%s():%d - returning %d\n", __func__, __LINE__, val);
+    return val;
+}
+
+bool ScMgr::epochAlreadyHasCertificate(const uint256& scId, int epochNumber, const CTxMemPool& pool, uint256& certHash)
+{
+    certHash = uint256();
+
+    // check the mempool first
+    for (auto it = pool.mapCertificate.begin(); it != pool.mapCertificate.end(); ++it)
+    {
+        const CScCertificate& mpCert = it->second.GetCertificate();
+
+        if ((mpCert.scId == scId) && (mpCert.epochNumber == epochNumber))
+        {
+            LogPrint("sc", "%s():%d - cert[%s] in mempool already refers to scId[%s] and epoch[%d]\n",
+                __func__, __LINE__, mpCert.GetHash().ToString(), scId.ToString(), epochNumber);
+            certHash = mpCert.GetHash();
+            return true;
+        }
+    }
+
+    // check db
+    ScInfo info;
+    if (!instance().getScInfo(scId, info) )
+    {
+        LogPrint("cert", "%s():%d - scId[%s] not found\n", __func__, __LINE__, scId.ToString() );
+        return false;
+    }
+
+    if (info.lastReceivedCertificateEpoch == epochNumber)
+    {
+        LogPrint("sc", "%s():%d - a cert for scId[%s] and epoch[%d] has already been received\n",
+            __func__, __LINE__, scId.ToString(), epochNumber);
+        return true;
+    }
```

Comment:
In case certificate is found "in db" its hash is not filled, unlike what happens in mempool.
Fix the incosistency